File: character_logic/balef.py
"""Balef battle and AI hooks."""

from __future__ import annotations

import copy
from typing import Any

from .common import floor_int, skill_key


CHARACTER_ID = "balef"
FLOW = "권류"
TRIO_KEYS = {skill_key(CHARACTER_ID, slot) for slot in range(3)}


def init_unique_state(fighter: Any, unique_names: set[str]) -> None:
    if FLOW in unique_names:
        fighter.counters[FLOW] = 0


def needs_battle_log(fighter: Any) -> bool:
    return True


def render_battle_log(battle: Any, fighter: Any, lines: list[str]) -> None:
    recent = [
        battle.display_action_name(fighter, action_key)
        for action_key in fighter.selected_attack_active_history[-3:]
    ]
    lines.append("공격 액티브 선택 기록: " + (" → ".join(recent) if recent else "없음"))
    trio = [skill_key(CHARACTER_ID, slot) for slot in range(3)]
    hits = [
        f"{battle.display_action_name(fighter, action_key)} {'완료' if action_key in fighter.hit_records else '미달성'}"
        for action_key in trio
    ]
    lines.append("삼위일권 명중: " + " / ".join(hits))


def on_make_choice(battle: Any, fighter: Any, action: Any, choice: Any) -> None:
    if action.is_active and action.is_attack:
        choice.prev_attack_active = fighter.selected_attack_active_history[-1] if fighter.selected_attack_active_history else None
        fighter.selected_attack_active_history.append(action.key)


def modify_priority(battle: Any, fighter: Any, action: Any, priority: int) -> int:
    if action.is_skill(CHARACTER_ID, 1):
        prev = fighter.selected_attack_active_history[-1] if fighter.selected_attack_active_history else None
        if prev == skill_key(CHARACTER_ID, 2):
            priority += 1
    return priority


def on_action_start(battle: Any, choice: Any) -> bool:
    actor = choice.actor
    action = choice.action
    if action.is_active and action.is_attack:
        prev = choice.prev_attack_active
        if prev and prev != action.key:
            actor.counters[FLOW] = int(actor.counters.get(FLOW, 0)) + 1
            print(f"권류가 1중첩 증가했다. 현재 {actor.counters[FLOW]}중첩")
    return False


def apply_condition_effects(battle: Any, choice: Any) -> bool | None:
    actor = choice.actor
    action = choice.action
    if action.is_skill(CHARACTER_ID, 2):
        if choice.prev_attack_active == skill_key(CHARACTER_ID, 0):
            choice.power = (choice.power or 0) + 10
            print("범권괴권 연계로 위력이 10 증가했다.")
    if action.is_skill(CHARACTER_ID, 3):
        if len(actor.selected_attack_active_history) < 2:
            return False
    return None


def attack_damage_multipliers(battle: Any, choice: Any) -> list[float]:
    actor = choice.actor
    target = battle.opponent(actor)
    action = choice.action
    multipliers = [1 + int(actor.counters.get(FLOW, 0)) * 0.05]
    if action.is_skill(CHARACTER_ID, 0):
        flow = int(actor.counters.get(FLOW, 0))
        if choice.prev_attack_active == skill_key(CHARACTER_ID, 2):
            multipliers.append(1.8)
        elif flow % 2 == 0:
            multipliers.append(1.4)
    if action.is_skill(CHARACTER_ID, 2):
        if target.defense_mult is not None:
            multipliers.append(3.0)
    return multipliers


def estimated_damage_multipliers(battle: Any, actor: Any, target: Any, action: Any) -> list[float]:
    projected_flow = _projected_flow(actor, action)
    multipliers = [1 + projected_flow * 0.05]
    prev = _prev_attack(actor)
    if action.is_skill(CHARACTER_ID, 0):
        if prev == skill_key(CHARACTER_ID, 2):
            multipliers.append(1.8)
        elif projected_flow % 2 == 0:
            multipliers.append(1.4)
    if action.is_skill(CHARACTER_ID, 2) and target.defense_mult is not None:
        multipliers.append(3.0)
    return multipliers


def estimated_power(battle: Any, actor: Any, target: Any, action: Any, power: int) -> int:
    if action.is_skill(CHARACTER_ID, 2) and _prev_attack(actor) == skill_key(CHARACTER_ID, 0):
        return power + 10
    return power


def on_hit_after_defense(battle: Any, choice: Any, total_damage: int) -> None:
    actor = choice.actor
    target = battle.opponent(actor)
    action = choice.action
    if action.is_active and action.is_attack:
        actor.hit_records.add(action.key)
        trio = {skill_key(CHARACTER_ID, 0), skill_key(CHARACTER_ID, 1), skill_key(CHARACTER_ID, 2)}
        if trio.issubset(actor.hit_records):
            fixed = floor_int(target.max_hp * 0.06)
            battle.fixed_damage(target, fixed, "삼위일권")
            actor.hit_records.difference_update(trio)
            if battle.game_over:
                return
    if action.is_skill(CHARACTER_ID, 0):
        if choice.prev_attack_active == skill_key(CHARACTER_ID, 1):
            battle.add_stat_effect(target, "atk", 0.9, 3, action.name)
            battle.add_stat_effect(actor, "atk", 1.1, 3, action.name)
    elif action.is_skill(CHARACTER_ID, 1):
        rate = 0.7 if choice.prev_attack_active == skill_key(CHARACTER_ID, 0) else 0.5
        amount = floor_int(total_damage * rate)
        reduced = battle.reduce_mp(target, amount, "흡성대권")
        battle.restore_mp(actor, reduced, "흡성대권")
    elif action.is_skill(CHARACTER_ID, 2):
        if choice.prev_attack_active == skill_key(CHARACTER_ID, 1):
            battle.heal(actor, floor_int(total_damage * 0.3), "관통마권")


def apply_non_attack_effects(battle: Any, choice: Any) -> bool:
    if not choice.action.is_skill(CHARACTER_ID, 3):
        return False
    actor = choice.actor
    history = actor.selected_attack_active_history
    if len(history) < 2:
        print("복제할 공격 액티브 기록이 부족하다.")
        return True
    actor.counters[FLOW] = int(actor.counters.get(FLOW, 0)) + 1
    first, second = history[-2], history[-1]
    original = actor.skill_by_key(first)
    if not original:
        return True
    copied = copy.copy(original)
    copied.mp = 0
    copied.power = floor_int((copied.power or 0) * 1.7)
    copied.accuracy = 100
    copied.priority = 0
    actor.selected_attack_active_history.append(copied.key)
    nested = choice.__class__(
        actor,
        copied,
        0,
        0,
        copied.power,
        copied.accuracy,
        prev_attack_active=second,
        copied_from=choice.action.name,
    )
    print(f"극의환권으로 {copied.name}을 복제해 즉시 처리한다.")
    battle.execute_action(nested, depth=1)
    return True


def would_condition_fail(battle: Any, actor: Any, target: Any, action: Any) -> bool | None:
    if action.is_skill(CHARACTER_ID, 3):
        return len(actor.selected_attack_active_history) < 2
    return None


def _prev_attack(actor: Any) -> str | None:
    return actor.selected_attack_active_history[-1] if actor.selected_attack_active_history else None


def _projected_flow(actor: Any, action: Any) -> int:
    flow = int(actor.counters.get(FLOW, 0))
    prev = _prev_attack(actor)
    if action.is_active and action.is_attack and prev and prev != action.key:
        flow += 1
    return flow


def _missing_trio(actor: Any) -> set[str]:
    return TRIO_KEYS.difference(actor.hit_records)


def _defense_read(battle: Any, target: Any, expected_damage: float) -> float:
    counts = battle.recent_kind_counts(target)
    value = counts["defense"] * 1.35 + counts["meditation"] * 0.25
    if expected_damage >= target.hp * 0.75:
        value += 1.0
    if expected_damage >= target.hp:
        value += 1.2
    if target.defense_streak >= 2:
        value -= 0.7
    return max(0.0, value)


def setup_value(battle: Any, actor: Any, target: Any, action: Any) -> float:
    if not action.is_skill(CHARACTER_ID, 3) or len(actor.selected_attack_active_history) < 2:
        return 0.0
    first, second = actor.selected_attack_active_history[-2], actor.selected_attack_active_history[-1]
    missing = _missing_trio(actor)
    value = 520 + int(actor.counters.get(FLOW, 0)) * 80
    if first in missing:
        value += 1450
        if len(missing) == 1:
            value += 1250
    if first != second:
        value += 420
    if TRIO_KEYS.issubset(actor.hit_records):
        value += floor_int(target.max_hp * 0.06) * 85
    return value


def ai_score(
    battle: Any,
    actor: Any,
    target: Any,
    action: Any,
    expected_damage: float,
    hit_rate: float,
) -> float:
    value = 0.0
    prev = _prev_attack(actor)
    missing = _missing_trio(actor)
    flow = int(actor.counters.get(FLOW, 0))
    projected_flow = _projected_flow(actor, action)
    defense_read = _defense_read(battle, target, expected_damage)
    missing_mp = max(0, 100 - actor.mp)

    if action.is_active and action.is_attack:
        if prev:
            value += 360 if action.key != prev else -260
        if TRIO_KEYS.issubset(actor.hit_records):
            value += floor_int(target.max_hp * 0.06) * 95
        elif action.key in missing:
            value += 520 * hit_rate
            if len(missing) == 1:
                value += 1500 * hit_rate
            elif len(missing) == 2:
                value += 520 * hit_rate
        elif missing:
            value -= 240

    if action.is_skill(CHARACTER_ID, 0):
        if prev == skill_key(CHARACTER_ID, 2):
            value += 820
        elif projected_flow % 2 == 0:
            value += 420
        if prev == skill_key(CHARACTER_ID, 1):
            value += 520
            if battle.estimate_best_incoming_damage(target, actor) >= actor.hp * 0.45:
                value += 380

    elif action.is_skill(CHARACTER_ID, 1):
        rate = 0.7 if prev == skill_key(CHARACTER_ID, 0) else 0.5
        expected_drain = min(target.mp, floor_int(max(0, expected_damage) * rate))
        useful_drain = min(missing_mp, expected_drain)
        if target.mp >= 25:
            value += expected_drain * 70
        if useful_drain >= 8:
            value += useful_drain * 140
        if actor.mp <= 55 and target.mp >= 45:
            value += 1300
        if prev == skill_key(CHARACTER_ID, 0):
            value += 620
            if actor.mp <= 60 and target.mp >= 45:
                value += 1300
        if prev == skill_key(CHARACTER_ID, 2):
            value += 360
        if actor.mp >= 86 and target.mp < 20:
            value -= 650

    elif action.is_skill(CHARACTER_ID, 2):
        if prev == skill_key(CHARACTER_ID, 0):
            value += 740
        if prev == skill_key(CHARACTER_ID, 1):
            value += 360 + min(700, max(0, 100 - actor.hp) * 8)
        if defense_read >= 2.2:
            value += 3400
        elif defense_read >= 1.2:
            value += 1400
        if target.defense_streak >= 2:
            value -= 520

    elif action.is_skill(CHARACTER_ID, 3):
        if len(actor.selected_attack_active_history) >= 2:
            first = actor.selected_attack_active_history[-2]
            if first in missing:
                value += 520
            elif missing:
                value -= 4200 if len(missing) == 1 else 1800
            if flow >= 3:
                value += 300
        if actor.mp < 58:
            value -= 360

    elif action.is_common_action("meditation"):
        if actor.mp < 39 and len(missing) <= 1:
            value += 460
        if actor.mp >= 90:
            value -= 520

    if action.is_common_action("normal_attack") and missing:
        value -= 280
    return value

File: web/battle-engine/character-logic/balef.js
"use strict";

const CHARACTER_ID = "balef";
const FLOW = "권의";

function skillKey(slot) {
  return `${CHARACTER_ID}:${slot}`;
}

function floorInt(value) {
  return Math.floor(value);
}

function history(fighter) {
  if (!Array.isArray(fighter.selectedAttackActiveHistory)) fighter.selectedAttackActiveHistory = [];
  return fighter.selectedAttackActiveHistory;
}

function previousAttack(fighter) {
  const selected = history(fighter);
  return selected.length ? selected[selected.length - 1] : null;
}

function projectedFlow(actor, action) {
  let flow = Number(actor.counters[FLOW] || 0);
  const previous = previousAttack(actor);
  if (action.isActive && action.isAttack && previous && previous !== action.key) flow += 1;
  return flow;
}

function missingTrio(actor) {
  return new Set([skillKey(0), skillKey(1), skillKey(2)].filter((key) => !actor.hitRecords.has(key)));
}

function defenseRead(battle, target, expectedDamage) {
  const counts = battle.recentKindCounts(target);
  let value = counts.defense * 1.35 + counts.meditation * 0.25;
  if (expectedDamage >= target.hp * 0.75) value += 1;
  if (expectedDamage >= target.hp) value += 1.2;
  if (target.defenseStreak >= 2) value -= 0.7;
  return Math.max(0, value);
}

function rawDamage(expectedDamage, hitRate) {
  return hitRate > 0 ? expectedDamage / hitRate : 0;
}

module.exports = {
  initUniqueState(fighter, uniqueNames) {
    if (uniqueNames.has(FLOW)) fighter.counters[FLOW] = 0;
    history(fighter);
  },

  needsBattleLog() {
    return true;
  },

  renderBattleLog(battle, fighter, lines) {
    const displayName = (key) => battle.findActionByInput(fighter, key)?.name || key;
    const recent = history(fighter).slice(-3).map(displayName);
    lines.push(`공격 액티브 선택 기록: ${recent.length ? recent.join(" -> ") : "없음"}`);
    lines.push(`삼위일권 명중: ${[0, 1, 2].map((slot) => `${displayName(skillKey(slot))} ${fighter.hitRecords.has(skillKey(slot)) ? "완료" : "미달성"}`).join(" / ")}`);
  },

  onMakeChoice(_battle, fighter, action, choice) {
    if (!action.isActive || !action.isAttack) return;
    choice.prevAttackActive = previousAttack(fighter);
    history(fighter).push(action.key);
  },

  modifyPriority(_battle, fighter, action, priority) {
    return action.isSkill(CHARACTER_ID, 1) && previousAttack(fighter) === skillKey(2) ? priority + 1 : priority;
  },

  onActionStart(battle, choice) {
    const actor = choice.actor;
    if (choice.action.isActive && choice.action.isAttack && choice.prevAttackActive && choice.prevAttackActive !== choice.action.key) {
      const before = Number(actor.counters[FLOW] || 0);
      actor.counters[FLOW] = before + 1;
      battle.logs.push(`${actor.name}의 ${FLOW} ${before} -> ${actor.counters[FLOW]}`);
    }
    return false;
  },

  applyConditionEffects(battle, choice) {
    if (choice.action.isSkill(CHARACTER_ID, 2) && choice.prevAttackActive === skillKey(0)) {
      choice.power = Number(choice.power || 0) + 10;
      battle.logs.push("범권괴권 연계로 위력이 10 증가했다.");
    }
    if (choice.action.isSkill(CHARACTER_ID, 3) && history(choice.actor).length < 2) return false;
    return true;
  },

  attackDamageMultipliers(battle, choice) {
    const actor = choice.actor;
    const action = choice.action;
    const flow = Number(actor.counters[FLOW] || 0);
    const multipliers = [1 + flow * 0.04];
    if (action.isSkill(CHARACTER_ID, 0)) {
      if (choice.prevAttackActive === skillKey(2)) multipliers.push(1.8);
      else if (flow % 2 === 0) multipliers.push(1.4);
    }
    if (action.isSkill(CHARACTER_ID, 2) && battle.opponent(actor).defenseMult !== null) multipliers.push(3);
    return multipliers;
  },

  estimatedDamageMultipliers(_battle, actor, target, action) {
    const flow = projectedFlow(actor, action);
    const previous = previousAttack(actor);
    const multipliers = [1 + flow * 0.04];
    if (action.isSkill(CHARACTER_ID, 0)) {
      if (previous === skillKey(2)) multipliers.push(1.8);
      else if (flow % 2 === 0) multipliers.push(1.4);
    }
    if (action.isSkill(CHARACTER_ID, 2) && target.defenseMult !== null) multipliers.push(3);
    return multipliers;
  },

  onHitAfterDefenseAsActor(battle, choice, totalDamage) {
    const actor = choice.actor;
    const target = battle.opponent(actor);
    const action = choice.action;
    const trio = [skillKey(0), skillKey(1), skillKey(2)];
    const trioWasReady = trio.every((key) => actor.hitRecords.has(key));
    if (action.isAttack && trioWasReady) {
      battle.fixedDamage(target, floorInt(target.maxHp * 0.05), "삼위일권", actor);
      for (const key of trio) actor.hitRecords.delete(key);
      if (battle.gameOver) return;
    } else if (action.isActive && action.isAttack && trio.includes(action.key)) {
      actor.hitRecords.add(action.key);
    }
    if (action.isSkill(CHARACTER_ID, 0) && choice.prevAttackActive === skillKey(1)) {
      battle.addStatEffect(target, "atk", 0.9, 3, action.name);
      battle.addStatEffect(actor, "atk", 1.1, 3, action.name);
    } else if (action.isSkill(CHARACTER_ID, 1)) {
      const rate = choice.prevAttackActive === skillKey(0) ? 0.7 : 0.5;
      const reduced = battle.reduceMp(target, floorInt(totalDamage * rate), "흡성대권");
      battle.restoreMp(actor, reduced, "흡성대권");
    } else if (action.isSkill(CHARACTER_ID, 2) && choice.prevAttackActive === skillKey(1)) {
      battle.heal(actor, floorInt(totalDamage * 0.3), "관통마권");
    }
  },

  applyNonAttackEffects(battle, choice) {
    if (!choice.action.isSkill(CHARACTER_ID, 3)) return false;
    const actor = choice.actor;
    const selected = history(actor);
    if (selected.length < 2) {
      battle.logs.push("복제할 공격 액티브 기록이 부족하다.");
      return true;
    }
    actor.counters[FLOW] = Number(actor.counters[FLOW] || 0) + 1;
    const first = selected[selected.length - 2];
    const second = selected[selected.length - 1];
    const original = battle.findActionByInput(actor, first);
    if (!original) return true;
    const copied = Object.assign(Object.create(Object.getPrototypeOf(original)), original, {
      mp: 0,
      power: floorInt(Number(original.power || 0) * 1.5),
      accuracy: 100,
      priority: 0,
    });
    selected.push(copied.key);
    battle.logs.push(`극의환권으로 ${copied.name}을 복제해 즉시 처리한다.`);
    battle.executeAction({
      actor,
      action: copied,
      cost: 0,
      priority: 0,
      power: copied.power,
      accuracy: copied.accuracy,
      hitCount: 1,
      guaranteedHit: false,
      selectedActionKey: copied.key,
      prevAttackActive: second,
      copiedFrom: choice.action.name,
      consumedMpExtra: 0,
      get totalCost() { return 0; },
    });
    return true;
  },

  estimatedPower(_battle, actor, _target, action, power) {
    return action.isSkill(CHARACTER_ID, 2) && previousAttack(actor) === skillKey(0) ? power + 10 : power;
  },

  setupValue(_battle, actor, target, action) {
    const selected = history(actor);
    if (!action.isSkill(CHARACTER_ID, 3) || selected.length < 2) return 0;
    const first = selected[selected.length - 2];
    const second = selected[selected.length - 1];
    const missing = missingTrio(actor);
    let value = 520 + Number(actor.counters[FLOW] || 0) * 80;
    if (missing.has(first)) {
      value += 1450;
      if (missing.size === 1) value += 1250;
    }
    if (first !== second) value += 420;
    if (missing.size === 0) value += floorInt(target.maxHp * 0.05) * 85;
    return value;
  },

  aiScore(battle, actor, target, action, expectedDamage, hitRate) {
    let value = 0;
    const damage = rawDamage(expectedDamage, hitRate);
    const previous = previousAttack(actor);
    const missing = missingTrio(actor);
    const flow = Number(actor.counters[FLOW] || 0);
    const flowAfterChoice = projectedFlow(actor, action);
    const read = defenseRead(battle, target, damage);
    const missingMp = Math.max(0, actor.maxMp - actor.mp);

    if (action.isAttack && missing.size === 0) {
      value += floorInt(target.maxHp * 0.05) * 95;
    }
    if (action.isActive && action.isAttack && missing.size > 0) {
      if (previous) value += action.key !== previous ? 360 : -260;
      if (missing.has(action.key)) {
        value += 520 * hitRate;
        if (missing.size === 1) value += 1500 * hitRate;
        else if (missing.size === 2) value += 520 * hitRate;
      } else value -= 240;
    }

    if (action.isSkill(CHARACTER_ID, 0)) {
      if (previous === skillKey(2)) value += 820;
      else if (flowAfterChoice % 2 === 0) value += 420;
      if (previous === skillKey(1)) {
        value += 520;
        if (battle.estimateBestIncomingDamage(target, actor) >= actor.hp * 0.45) value += 380;
      }
    } else if (action.isSkill(CHARACTER_ID, 1)) {
      const rate = previous === skillKey(0) ? 0.7 : 0.5;
      const expectedDrain = Math.min(target.mp, floorInt(Math.max(0, damage) * rate));
      const usefulDrain = Math.min(missingMp, expectedDrain);
      if (target.mp >= 25) value += expectedDrain * 70;
      if (usefulDrain >= 8) value += usefulDrain * 140;
      if (actor.mp <= 55 && target.mp >= 45) value += 1300;
      if (previous === skillKey(0)) {
        value += 620;
        if (actor.mp <= 60 && target.mp >= 45) value += 1300;
      }
      if (previous === skillKey(2)) value += 360;
      if (actor.mp >= 86 && target.mp < 20) value -= 650;
    } else if (action.isSkill(CHARACTER_ID, 2)) {
      if (previous === skillKey(0)) value += 740;
      if (previous === skillKey(1)) value += 360 + Math.min(700, Math.max(0, 100 - actor.hp) * 8);
      if (read >= 2.2) value += 3400;
      else if (read >= 1.2) value += 1400;
      if (target.defenseStreak >= 2) value -= 520;
    } else if (action.isSkill(CHARACTER_ID, 3)) {
      const selected = history(actor);
      if (selected.length >= 2) {
        const first = selected[selected.length - 2];
        if (missing.has(first)) value += 520;
        else if (missing.size) value -= missing.size === 1 ? 4200 : 1800;
        if (flow >= 3) value += 300;
      }
      if (actor.mp < 58) value -= 360;
    } else if (action.isCommonAction("meditation")) {
      if (actor.mp < 39 && missing.size <= 1) value += 460;
      if (actor.mp >= 90) value -= 520;
    }
    if (action.isCommonAction("normal_attack") && missing.size) value -= 280;
    return value;
  },

  wouldConditionFail(_battle, actor, _target, action) {
    return action.isSkill(CHARACTER_ID, 3) ? history(actor).length < 2 : null;
  },
};
